### ISSUE TITLE
Support chunksize in I/O functionalities

### ISSUE THREAD

**Issue by devin-petersohn** [modin-project/modin#923]
### Describe the problem

Chunksize defaults to pandas in all cases. Ideally, we will be able to support it. The functionality should not require deep analysis, we can just mask the result of reading with `chunksize=None`.

Relevant conversation in #909.

### Source code / logs

Proposed solution for `read_sql`:
```python
_, _, _, kwargs = inspect.getargvalues(inspect.currentframe())
kwargs["chunksize"] = None
df = DataFrame(query_compiler=BaseFactory.read_sql(**kwargs))
if chunksize is not None:
    num_chunks = len(df) // chunksize if len(df) % chunksize == 0 else len(df) // chunksize + 1
    return (df.iloc[i * chunksize: (i + 1) * chunksize] for i in num_chunks)
else:
    return df
```

**Comment by manesioz** [comment]
I have a (probably dumb) question, but why do you write `kwargs["chunksize"] = None` before checking if its `None`? Wouldn't that ensure that it is always `None`? The current implementation checks if it is `None` like this: 
```python
if kwargs.get("chunksize") is not None:
        #default to pandas
```

**Comment by devin-petersohn** [comment]
`chunksize` is the parameter passed in above, `kwargs` is a dictionary collection of the values provided to the function. They are separate variables.

We set `kwargs["cunksize"] = None` because the internal implementation does not support it. Those are just the args sent to the internal implementation. What's written in my original comment is equivalent to this (just fewer lines):

```python
_, _, _, kwargs = inspect.getargvalues(inspect.currentframe())
if kwargs.get("chunksize", None) is not None:
    kwargs["chunksize"] = None
    df = DataFrame(query_compiler=BaseFactory.read_sql(**kwargs))
    num_chunks = len(df) // chunksize if len(df) % chunksize == 0 else len(df) // chunksize + 1
    return (df.iloc[i * chunksize: (i + 1) * chunksize] for i in num_chunks)
else:
    return DataFrame(query_compiler=BaseFactory.read_sql(**kwargs))
```

**Comment by manesioz** [comment]
Thank you for the clarification! Another small thing I was wondering about before making a PR: 
In your implementation above it seems like if the `chunksize` is not null you call `read_sql()` naively (without any `chunksize`) and then partition it after - does that defeat the purpose of having a `chunksize` in the first place? 
I say this because if a particular query returns a very large result, we internally request and store it in memory anyway. Would it be more performant to yield the result (we would not have to store it all in memory if it returns a generator)? Or is there a way to partition the results at a lower level? 
Sorry if this is not making sense, but I suppose I am wondering how this implementation results in any performance improvements.

**Comment by devin-petersohn** [comment]
You are right that there would be a performance impact, @eavidan had an implementation here that followed the `yield` model: https://github.com/modin-project/modin/issues/909#issuecomment-568479921

The motivation for doing it all at once was to avoid paying the fixed distributed computing overheads multiple times and for smaller datasets. There's probably some benchmarking necessary to see what the pro/cons are. It's not obvious which is better (at least to me).

**Comment by manesioz** [comment]
Interesting. Out of curiosity, how do you normally approach these trade-offs that are present in distributed systems? Do you provide the user with as many options are possible and let them make the decision of what to use/when, or do you choose the most common/robust approach? 
From a design perspective I guess I am wondering if you want to include a `yield`-based approach or simply an in-memory approach, or both. Let me know what you think and if a PR should be made :)

**Comment by devin-petersohn** [comment]
If it changes the semantics/syntax, it goes in `experimental`. The `yield`-based approach probably fits better with the semantics/usage of `pandas`.

@eavidan Any opinion on this?

**Comment by eavidan** [comment]
PRs are always welcomed @manesioz :)

I believe the `yield`-based approach fits the chunking principle of `pandas` better. For instance, querying in small chucks should result in a quick response from the database during the first iteration. It is expected that each iteration starts with a delay for getting its respectful chunk. Waiting a long time for the first iteration to start is not an expected behavior with a small `chunksize`. Moreover, the user may decide to stop iterating at some point, and this will not improve performance if everything was loaded into memory already.

**Comment by devin-petersohn** [comment]
I agree @eavidan. In that case, let's go with the `yield`.

**Comment by manesioz** [comment]
In @eavidan 's generator-based approach, is there an issue with this line?  
```python
kwargs["sql"] = "SELECT * FROM ({}) LIMIT {} OFFSET {}".format(sql, chunksize, offset)
```
Since you get no guarantees from your RDBMS that your data will come back in a certain order -- or even in a consistent order -- unless you query your data with an `ORDER BY` clause, is it possible that the `OFFSET` will _not_ select perfectly disjointed sub-sets of rows? Or am I looking at this wrong?

**Comment by eavidan** [comment]
That's a good point @manesioz. My assumption that the order is arbitrary but _constant_ may not hold in all cases. We should probably sort by some row unique id. What do you think about using the ROWID column?
```py
kwargs["sql"] = "SELECT * FROM ({}) ORDER BY ROWID LIMIT {} OFFSET {}".format(sql, chunksize, offset)
```
We would need to check if the user has an `ORDER BY` in their query first, since there is no need to add our own unless the user did not define one of his own.
Another thing that comes to mind is the fact that rows can be added to the table while the user iterates it. I don't believe it will be an issue, since we added the ROWID, and these lines will only be fetched at the end. 
what do you think?

**Comment by devin-petersohn** [comment]
I've tested on multiple implementations (Postgres, MySQL, sqlite), and the order is the same in each of those implementation across the workers that connect to it.

It is true that the data model does not guarantee an order, but implementations of relational systems generally do not change the order of the data because they have indexes and other structures built against the current order. If there are updates to the database (i.e. new rows) during the query, depending on the implementation it may or may not cause an issue. The only other time it should become a problem would be if someone is altering the configuration of a given database at the same time as a query is executing from the different workers.

We can do two things:

1.) Warn the user when using `read_sql` that updates to the database during query execution may yield non-deterministic results
2.) Figure out a way to do this with cursors and just scroll the cursor through the database rows.

I haven't mentally mapped through the cursor approach, so I'm not sure how well it would work. There may also be issues with SQL implementations and language differences, I am not sure.

**Comment by eavidan** [comment]
I agree @devin-petersohn. I haven't encountered any ordering inconsistencies before so I think we should handle these issues if and when they are reported. So, I tend to go with the first approach.
Regarding the cursors, it would be nice if we could use `fetchmany()` across different partitions, but I believe this will not be possible, due to the state this object holds, to serialize it.

**Comment by manesioz** [comment]
> What do you think about using the `ROWID` column?

This is specific to certain database implementations and should probably not be hard-coded. 

> We would need to check if the user has an `ORDER BY` in their query first, since there is no need to add our own unless the user did not define one of his own.

I am under the impression that this is not necessarily true, and that to ensure order there must be an `ORDER BY` on the outside. 

> Another thing that comes to mind is the fact that rows can be added to the table while the user iterates it

Good point, as @devin-petersohn mentioned this could be a problem. I suppose he's right, there is usually not an issue with ordering regardless, but this is not guaranteed and I'd prefer not to just warn the user that it may have non-deterministic results. 

I've glanced at panda's implementation and like you mentioned, they use a `yield`-based approach with `fetchmany(chunksize)`. Could you explain what you mean when you say:

> it would be nice if we could use `fetchmany()` across different partitions, but I believe this will not be possible, due to the state this object holds, to serialize it.

I am not too familiar with the `modin` codebase and architecture, so which part would be serialized and distributed? 
Assuming the function that uses `fetchmany()` looks like this: 

```python
def _query_iterator(
    result, chunksize, columns, index_col=None, coerce_float=True, parse_dates=None
):
    """Return generator through chunked result set"""

    while True:
        data = result.fetchmany(chunksize)
        if not data:
            break
        else:
            yield _wrap_result(
                data,
                columns,
                index_col=index_col,
                coerce_float=coerce_float,
                parse_dates=parse_dates,
            )
```

**Comment by devin-petersohn** [comment]
Serialization happens on any distributed operation. Generators are generally not picklable (https://bugs.python.org/issue1092962) so we could not return any generator from a distributed computation. The computation would have to be done without partitioning if it were to return a generator, which is the current approach: To allow pandas to query the database and then distribute the data after pandas has finished building the DataFrame.

**Comment by manesioz** [comment]
Ok, so is that approach not performant enough? How should this PR be approached, if at all?

**Comment by devin-petersohn** [comment]
Why don't we try to implement with `fetchmany`. It may not work, but it will be worth trying. Otherwise, the right approach would be to use @eavidan's approach and throw a warning about the order.
